src/VisualStudio/CSharp/Impl/SemanticSearch/SemanticSearchToolWindowImpl.cs

```diff
@@ -0,0 +1,470 @@
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.CSharp;
+
+using TextSpan = Microsoft.CodeAnalysis.Text.TextSpan;
+
+[Shared]
+[Export(typeof(ISemanticSearchWorkspaceHost))]
+[Export(typeof(SemanticSearchToolWindowImpl))]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class SemanticSearchToolWindowImpl(
+    IHostWorkspaceProvider hostWorkspaceProvider,
+    IThreadingContext threadingContext,
+    ITextEditorFactoryService textEditorFactory,
+    ITextDocumentFactoryService textDocumentFactory,
+    IContentTypeRegistryService contentTypeRegistry,
+    IVsEditorAdaptersFactoryService vsEditorAdaptersFactoryService,
+    IAsynchronousOperationListenerProvider listenerProvider,
+    IGlobalOptionService globalOptions,
+    VisualStudioWorkspace workspace,
+    IStreamingFindUsagesPresenter resultsPresenter,
+    SVsServiceProvider serviceProvider) : ISemanticSearchWorkspaceHost, OptionsProvider<ClassificationOptions>
+{
+    private const int ToolBarHeight = 26;
+    private const int ToolBarButtonSize = 20;
+
+    private static readonly Lazy<ControlTemplate> s_buttonTemplate = new(CreateButtonTemplate);
+
+    private readonly IVsUIShell _vsUIShell = (IVsUIShell)serviceProvider.GetService(typeof(SVsUIShell));
+    private readonly IContentType _contentType = contentTypeRegistry.GetContentType(ContentTypeNames.CSharpContentType);
+    private readonly IAsynchronousOperationListener _asyncListener = listenerProvider.GetListener(FeatureAttribute.SemanticSearch);
+
+    private readonly Lazy<SemanticSearchEditorWorkspace> _semanticSearchWorkspace
+        = new(() => new SemanticSearchEditorWorkspace(
+            hostWorkspaceProvider.Workspace.Services.HostServices,
+            CSharpSemanticSearchUtilities.Configuration,
+            threadingContext,
+            listenerProvider));
+
+    // Access on UI thread only:
+    private CancellationTokenSource? _pendingExecutionCancellationSource;
+    private Button? _executeButton;
+    private Button? _cancelButton;
+    private IWpfTextView? _textView;
+    private ITextBuffer? _textBuffer;
+
+    public async Task<FrameworkElement> InitializeAsync(CancellationToken cancellationToken)
+    {
+        // TODO: replace with XAML once we can represent the editor as a XAML element
+        // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1927626
+
+        // TODO: Add toolbar and convert Execute and Cancel buttons to commands.
+        // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1978331
```

Comment:
Can you not do it like this?
https://github.com/dotnet/roslyn/blob/02069cef3ee4f08b37a1b5f2371c26041c41a5bd/src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineView.xaml.cs#L63-L69

Reply:
IDK. Maybe? I would rather not mess with the UI anymore. The current implementation works well and it's all just workaround until Gladstone supports toolbars in tool windows natively. At that point we'll be able to delete all this.

Reply:
📝 This may be a blocking issue due to accessibility requirements.

Reply:
Which accessibility issues? I tested theming and it works well. The buttons have tool tip on hover. I was going to follow up with adding commands so you can trigger the Execute and Cancel operations with keyboard. Anything else?

I'll try the above but Gladstone Tool Window isn't exactly normal tool window so some stuff doesn't work.

Reply:
Added command filter with key bindings similar to Interactive Window: Ctrl+Enter to run query and Escape to cancel running query.

Reply:
Adding automation properties for the buttons. Narrator now works.